scipy/optimize/_constraints.py

```diff
@@ -133,25 +133,75 @@ class LinearConstraint:
     ----------
     A : {array_like, sparse matrix}, shape (m, n)
         Matrix defining the constraint.
-    lb, ub : array_like
-        Lower and upper bounds on the constraint. Each array must have the
+    lb, ub : array_like, optional
+        Lower and upper limits on the constraint. Each array must have the
         shape (m,) or be a scalar, in the latter case a bound will be the same
         for all components of the constraint. Use ``np.inf`` with an
         appropriate sign to specify a one-sided constraint.
         Set components of `lb` and `ub` equal to represent an equality
         constraint. Note that you can mix constraints of different types:
         interval, one-sided or equality, by setting different components of
-        `lb` and `ub` as  necessary.
+        `lb` and `ub` as  necessary. Defaults to ``lb = -np.inf``
+        and ``ub = np.inf`` (no limits).
     keep_feasible : array_like of bool, optional
         Whether to keep the constraint components feasible throughout
         iterations. A single value set this property for all components.
         Default is False. Has no effect for equality constraints.
     """
-    def __init__(self, A, lb, ub, keep_feasible=False):
-        self.A = A
-        self.lb = lb
-        self.ub = ub
-        self.keep_feasible = keep_feasible
+    def _input_validation(self):
+        if self.A.ndim != 2:
+            message = "`A` must have exactly two dimensions."
+            raise ValueError(message)
+
+        try:
+            shape = self.A.shape[0:1]
+            self.lb = np.broadcast_to(self.lb, shape)
+            self.ub = np.broadcast_to(self.ub, shape)
+            self.keep_feasible = np.broadcast_to(self.keep_feasible, shape)
```

Comment:
I can see why this is here (because asserting "broadcast-ability" seems to necessitate actually doing the broadcast) but will quickly note that `_input_validation` is mutating class members, not strictly validating.  The `__init__` function below (line 166) also contains methods that mutate as a setup step to this function.  I would prefer to see all mutation happen in `__init__`, but I'm open to the feedback that this is a generally standard separation of functionality in this code base or there's a strong opinion the other way

Reply:
I have gotten in the habit of moving input validation and standardization into a function/method that is separate from wherever it is actually used. It is not standard, but I think it is usually a good idea.

The most important reason is to move this sort of tedious, boring code away from the important logic to make the code easier to read. It also makes sure that IV/standardization is done _before_ getting into the logic, instead of being interspersed with it (or not done at all). Also, by giving this part of the code a name (`_input_validation`, or maybe more appropriately, (`_input_validation_and_standardization`), it emphasizes to future contributors that there is input validation/standardization going on, so they can expect the variables to be in a certain format when they're working with them in the logic part of the code. (Perhaps it also reminds them to consider input validation/standardization, in their own code, as this is often neglected.) 
 
In functions/methods where there is more going on after input validation/standardization, I suppose it makes more sense than here, where it's essentially all that is going on in `__init__`. I didn't really think about it; it's a reflex now.

As to why there is some of it in the main part of `__init__` and most of it in this separate function: the statements that are still in the main part of `__init__` were basically already there, and I didn't move them. Also, I suppose they make it so that I don't have to pass a bunch of arguments into `_input_validation`, and by being there, the reader can still see at a glance what instance variables there are. Exactly what is done in `_input_validation` is less important to the rest of the class.